### ISSUE TITLE
Tab crashes when viewing songs

### ISSUE THREAD

**Issue by pedrogarcia** [zachomedia/apple-music-webplayer#29]
About 1.29GB of unique album art data ([assuming duplicate data is not served by the MusicKit.js API](https://github.com/zachomedia/apple-music-webplayer/issues/26)) is loaded for the around 10.000 songs in my Music Library. 

The tab crashes on Chrome when about a third if data is loaded, when viewing the list of songs. The app may be hitting memory limits (it crashes at around 4GB) as this is reproduced in a computer with 8GB of RAM.

**Comment by zachomedia** [comment]
I've been thinking about this on and off recently.. I like having the artwork appear on the songs list, but for a large number of songs it becomes a problem.

There are two fixes, neither of which are ideal: not show artwork or paginate the page so only `x` songs are visible at a time.

I'm open to other possible solutions too.

**Comment by pedrogarcia** [comment]
I'm not a web developer, so my suggestions may not be that useful. 

Well, to be fair, the problem is really that you're being served 10.000 different copies of High-resolution, 1400x1400 images. Is there anyway to cache some data to disk?

I like having the artwork appearing on the songs list as well. 

GPM, for example, deals with this problem by showing only low resolution images (for the small thumbnail), and only fetching them when they are in scrolled into view. 

Besides pagination and not showing artwork, one way I can think is to only load the image for the first track of an album, and then, every other track on the album should show that image instead. 

Another, is to group all the songs of an album together, instead of alphabetical order and only show 1 copy of the artwork. 

Perhaps you could paginate with infinite scrolling? Or paginate grouping by letter? Or only show artwork when there is less than x number of songs in Library?

**Comment by zachomedia** [comment]
So the hardest part is that I have no control over the images (where they comes from, their actual size, the cache, etc.) because that's all controlled by Apple. I only ask MusicKit JS for a URL to an image for `x` pixels in size.

Now, I did some investigation just now - I'm seeing that apple is sending 40x40px images for the songs list and they're about ~1-2KB each. However, their are songs that have large PNGs for their artwork, with sizes of ~150-200KB each (now, this relates to #26, but I don't see duplication for the 40x40px images but I do for the large PNG images).

Thanks for some additional ideas. I'll look at prototyping some solutions.

**Comment by pedrogarcia** [comment]
> I only ask MusicKit JS for a URL to an image for x pixels in size.

> Now, I did some investigation just now - I'm seeing that apple is sending 40x40px images for the songs list and they're about ~1-2KB each.

Is this implemented yet? When I right click a song, and open the image in a new tab, it shows the full artwork. 

You should test with an uploaded song (with custom artwork) as well, because I only saw the 40x40 image for tracks added from Apple Music.

**Comment by zachomedia** [comment]
Yeah, tracks added from Apple Music give me actual images at 40x40px. But music that I'm guessing was matched / uploaded from my library does not - it gives me the full size images like you describe. It seems to affect artwork that gets loaded from Amazon S3 rather than mzstatic.com.

I'm in the process of writing this into a bug report with Apple.

**Comment by zachomedia** [comment]
Radar 42698851
